server/ladder_service/veto_system.py

```diff
@@ -0,0 +1,213 @@
+
+        if adjusted_vetoes != self._vetoes:
+            self._vetoes = adjusted_vetoes
+            return veto_datas
+        return None
+
+    def _is_valid_vetoes(self, vetoes: dict) -> bool:
+        return (
+            isinstance(vetoes, dict)
+            and all(
+                isinstance(k, int)
+                and isinstance(v, dict)
+                and all(
+                    isinstance(mk, int) and isinstance(mv, int) and mv >= 0
+                    for mk, mv in v.items()
+                )
+                for k, v in vetoes.items()
+            )
+        )
+
+    def get_correct_vetoes_for_bracket(self, new_bracket_vetoes: dict[MapPoolMapVersionId, VetoTokensApplied], map_ids: list[MapPoolMapVersionId], total_tokens: int, max_per_map: int) -> dict[MapPoolMapVersionId, VetoTokensApplied]:
+        adjusted_vetoes = {}
+        tokens_sum = 0
+        for map_id in map_ids:
+            tokens = new_bracket_vetoes.get(map_id, 0)
+            tokens_applied = self.cap_tokens(tokens, tokens_sum, total_tokens, max_per_map)
+            if tokens_applied > 0:
+                adjusted_vetoes[map_id] = tokens_applied
+                tokens_sum += tokens_applied
+        return adjusted_vetoes
+
+    def cap_tokens(self, tokens: int, tokens_sum: int, total_tokens: int, max_per_map: int) -> int:
+        tokens_applied = max(tokens, 0)
+        if tokens_sum + tokens_applied > total_tokens:
+            tokens_applied = total_tokens - tokens_sum
+        if max_per_map > 0 and tokens_applied > max_per_map:
+            tokens_applied = max_per_map
+        return tokens_applied
+
+    def build_veto_data(self, bracket_id: BracketID, vetoes: dict[MapPoolMapVersionId, VetoTokensApplied]) -> list[dict]:
+        """Builds veto data for sending to the client."""
+        return [
+            {
+                "map_pool_map_version_id": map_id,
+                "veto_tokens_applied": tokens,
+                "matchmaker_queue_map_pool_id": bracket_id
+            }
+            for map_id, tokens in vetoes.items()
+        ]
+
+    def get_vetoes_for_bracket(self, bracket_id: BracketID) -> dict[MapPoolMapVersionId, VetoTokensApplied]:
+        return self._vetoes.get(bracket_id, {})
+
+
+@with_logger
+class VetoSystem:
+    _logger: ClassVar[logging.Logger] = logging.getLogger(__name__)
+    _max_stream_count = 25
+    pools_veto_data: ClassVar[list[MatchmakerQueueMapPoolVetoData]] = []
+
+    @staticmethod
+    async def apply_vetoes_to_player(player: Player, new_vetoes: Optional[dict[BracketID, dict[MapPoolMapVersionId, VetoTokensApplied]]] = None) -> None:
+        """Applies vetoes for a player and sends a message if changes occur."""
+        veto_datas = player.vetoes.apply_vetoes(new_vetoes)
+        if veto_datas:
+            try:
+                await player.send_message({
+                    "command": "vetoes_changed",
+                    "vetoesData": veto_datas
```

Comment:
BTW our message protocol uses snake case.
```suggestion
                    "vetoes_data": veto_datas
```